entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -89,14 +91,29 @@ public void execute(EntityQueryRequest request, StreamObserver<ResultSetChunk> r
         .convertToEDSQuery(request, attrNameToEDSAttrMap.get(request.getEntityType()));
     Iterator<Document> documentIterator = entitiesCollection.search(
         DocStoreConverter.transform(tenantId.get(), query));
-    List<Entity> entities = convertDocsToEntities(documentIterator);
-
-    //Build result
-    //TODO : chunk response. For now sending everything in one chunk
-    responseObserver.onNext(convertEntitiesToResultSetChunk(
-        entities,
-        request.getSelectionList(),
-        attrNameToEDSAttrMap.get(request.getEntityType())));
+
+    boolean isFirstEntity = true;
+    while (documentIterator.hasNext()) {
+      Optional<Entity> entity = DOCUMENT_PARSER.parseOrLog(documentIterator.next(), Entity.newBuilder());
+      if (entity.isPresent()) {
+        Row row = convertToEntityQueryResult(entity.get(), request.getSelectionList(), attrNameToEDSAttrMap.get(request.getEntityType()));
+        ResultSetChunk.Builder resultBuilder = ResultSetChunk.newBuilder();
+        // Set metadata for first entity
+        if (isFirstEntity) {
+          resultBuilder.setResultSetMetadata(ResultSetMetadata.newBuilder()
+              .addAllColumnMetadata(
+                  () -> request.getSelectionList().stream().map(Expression::getColumnIdentifier)
+                      .map(
+                          ColumnIdentifier::getColumnName)
+                      .map(s -> ColumnMetadata.newBuilder().setColumnName(s).build()).iterator())
+              .build());
+          isFirstEntity = false;
+        }
+        //Build data
+        resultBuilder.addRow(row);
```

Comment:
Looks like, with chunking, we need to increment chunk_id and also need to take care of last chunk as per API - https://github.com/hypertrace/entity-service/blob/4025041f432818bdbfb0b6123cdd2fa7bedeadff/entity-service-api/src/main/proto/org/hypertrace/entity/query/service/v1/response.proto

Query-service also seems to have a similar protocol (ResultSetChunk) - https://github.com/hypertrace/query-service/blob/466b05904139ff2f4c1de8de46597a03599d2ec3/query-service-api/src/main/proto/response.proto

You can also see the impl of chunk_id and is_last_chunk here as a reference - https://github.com/hypertrace/query-service/blob/466b05904139ff2f4c1de8de46597a03599d2ec3/query-service-impl/src/main/java/org/hypertrace/core/query/service/RowChunkingOperator.java

Reply:
Updated